lucetc/src/timing.rs

```diff
@@ -0,0 +1,68 @@
+use crate::timing;
+use serde::Serialize;
+use std::cell::RefCell;
+use std::fmt;
+use std::mem;
+use std::time::Duration;
+
+thread_local! {
+    static OUTPUT_TIME: RefCell<Duration> = RefCell::new(Duration::default());
+}
+
+pub(crate) fn record_output_time(duration: Duration) {
+    OUTPUT_TIME.with(|rc| *rc.borrow_mut() = duration);
+}
+
+pub fn take_output_time() -> Duration {
+    OUTPUT_TIME.with(|rc| mem::replace(&mut *rc.borrow_mut(), Duration::default()))
+}
+
+#[derive(Serialize)]
+pub struct TimingInfo {
+    pass_times: Vec<String>,
+}
+
+impl TimingInfo {
+    pub fn collect() -> Self {
+        // `cranelift_codegen::timing::PassTimes` has hidden members at the moment
+        // so the best we can do consistently without deep sins like transmutes is to just split
+        // some strings.
+        let mut pass_times: Vec<String> = vec![];
+        let cranelift_time_text = cranelift_codegen::timing::take_current().to_string();
+        // skip the header text from Cranelift's `Display`, then take until we hit the end (also
+        // "======= ======= ==...")
+        for pass in cranelift_time_text
```

Comment:
this is some real nastyness but I don't see a way to record duration/times without stringifying and splitting. I'll follow up in cranelift and see about exposing this in a more friendly form.